docs: Pulumi comparison (#462)
nitrictech/docs: src/pages/faq/comparison/pulumi.mdx

export const description = 'Comparison to Pulumi'

# Pulumi vs. Nitric

> > NOTE: Nitric is not designed to replace IaC tools like Pulumi, but instead introduces a method of bringing developer self service for infrastructure directly into the developer application (Nitric's default deployment engines are built with Pulumi).
> > Nitric can be augmented through use of tools like Pulumi and even be fully customized using such tools see [Custom Providers]() for more details

Reviewer: ```suggestion
<Note>Nitric is not designed to replace IaC tools like Pulumi, but instead introduces a method of bringing developer self service for infrastructure directly into the developer application (Nitric's default deployment engines are built with Pulumi). Nitric can be augmented through use of tools like Pulumi and even be fully customized using such tools see [Custom Providers]() for more details</Note>
```

Author: done

Nitric is a framework designed to aid developers in building full cloud applications, including declaring their infrastructure and application code in one place. Pulumi is an Infrastructure as Code tool that enables developers to define infrastructure using traditional programming languages or YAML (Cuelang). The main differences between these are:

1. Nitric is cloud-agnostic, code that is written using nitric constructs can be deployed to any cloud. Pulumi supports many clouds, infrastructure declarations are explicitly defined for the provider that a resource is provided by.

1. Nitric defines not only the infrastructure but how it is interacted with at runtime, so infrastructure can be automatically inferred from application code to ensure best practice deployments and least privilege security.

> 1. Nitric uses Pulumi under the hood for the Out of the Box providers. By extensions Pulumi is a great option for extending infrastructure that is deployed by nitric, or can even be used to deploy Nitric applications to clouds Nitric does not support by default.

Reviewer 2: ```suggestion
1. Nitric uses Pulumi under the hood for the Out of the Box providers. Pulumi is a great option for extending infrastructure deployed by Nitric or for deploying Nitric applications to clouds not supported by Nitric by default.
```

> 1. Nitric provides tools for locally simulating a cloud environment (using the nitric CLI), to allow application to be tested locally. Pulumi programs can be unit tested but applications written to be deployed will need a separate solution for local development such as LocalStack.

Reviewer 2: ```suggestion
1. Nitric provides tools for locally simulating a cloud environment (using the nitric CLI), to allow applications to be tested locally. Pulumi programs can be unit tested but applications written to be deployed will need a separate solution for local development such as LocalStack.
```

        },
      },
    ],
  }),
})

const policy = new aws.iam.RolePolicy('lambdaPolicy', {
  role: role,
  policy: JSON.stringify({
    Version: '2012-10-17',
    Statement: [
      {
        Action: [
          'logs:CreateLogGroup',
          'logs:CreateLogStream',
          'logs:PutLogEvents',
        ],
        Effect: 'Allow',
        Resource: 'arn:aws:logs:*:*:*',
      },
      {
        Action: ['s3:PutObject'],
        Effect: 'Allow',
        Resource: `${bucket.arn}/*`,
      },
    ],
  }),
})

const lambdaFunction = new aws.lambda.Function('s3UploaderLambda', {
  runtime: 'nodejs14.x',
  code: new pulumi.asset.AssetArchive({
    '.': new pulumi.asset.FileArchive('./'),
  }),
  timeout: 10,
  handler: 'index.handler',
  role: role.arn,
  environment: {
    variables: {
      BUCKET_NAME: bucket.bucket,
    },
  },
})

export const bucketName = bucket.bucket
export const lambdaFunctionName = lambdaFunction.name
```

```javascript {{title: 'Pulumi', label: 'src/index.js'}}
const AWS = require('aws-sdk')

exports.handler = async (event) => {
  const {
    s3: { object },
  } = event
  console.log(object.key, 'was created')
}
```

</CodeGroup>

> > In the pulumi example we get control over absolutely every facet of our infrastructure which is great, but for the majority of use cases a repeatable pattern of deployment is enough and scales much better than applying fine grained configuration every time we need to define a new deployment (with Pulumi this would typically be done by sharing infrastructure as libraries e.g. Pulumi Crosswalks).

Reviewer: ```suggestion
<Note>In the pulumi example we get control over absolutely every facet of our infrastructure which is great, but for the majority of use cases a repeatable pattern of deployment is enough and scales much better than applying fine grained configuration every time we need to define a new deployment (with Pulumi this would typically be done by sharing infrastructure as libraries e.g. Pulumi Crosswalks).</Note>
```

Author: done

The below table contains the main differences that you can see in the code examples, and also some that cannot fit in such a small app, but we still would like you to know about :)

| Feature              | Nitric                                                                     | Pulumi                                                 |
| -------------------- | -------------------------------------------------------------------------- | ------------------------------------------------------ |
| Language             | Your choice                                                                | YAML + your choice                                     |
| Lines of code        | 7                                                                          | 69                                                     |

> | Cloud Infrasutrcture | Inferred                                                                   | Explicit                                               |

Reviewer 2: ```suggestion
| Cloud Infrastructure | Inferred                                                                   | Explicit                                               |
```

export const description = 'Comparison to Pulumi'

# Pulumi vs. Nitric

<Note>

>   NOTE: Nitric is not designed to replace IaC tools like Pulumi, but instead

Reviewer: ```suggestion
  Nitric is not designed to replace IaC tools like Pulumi, but instead
```
